ufmt/core.py

```diff
@@ -2,118 +2,178 @@
 # Licensed under the MIT license
 
 import logging
-from dataclasses import dataclass
 from functools import partial
 from pathlib import Path
 from typing import List, Optional
+from warnings import warn
 
-from black import (
-    decode_bytes,
-    find_pyproject_toml,
-    format_file_contents,
-    Mode,
-    NothingChanged,
-    parse_pyproject_toml,
-    TargetVersion,
-)
+from black import format_file_contents, NothingChanged
 from moreorless.click import unified_diff
 from trailrunner import Trailrunner
-from usort.config import Config as UsortConfig
-from usort.sorting import usort_string
+from usort import usort
 
 from .config import ufmt_config
+from .types import (
+    BlackConfig,
+    BlackConfigFactory,
+    Encoding,
+    FileContent,
+    Result,
+    UsortConfig,
+    UsortConfigFactory,
+)
+from .util import make_black_config, read_file, write_file
 
 LOG = logging.getLogger(__name__)
 
 
-@dataclass
-class Result:
-    path: Path
-    changed: bool = False
-    written: bool = False
-    diff: Optional[str] = None
-
-
-def ufmt_string(
+def ufmt_bytes(
     path: Path,
-    content: str,
+    content: FileContent,
+    *,
+    black_config: BlackConfig,
     usort_config: UsortConfig,
-    black_config: Optional[Mode] = None,
-) -> str:
-    content = usort_string(content, usort_config, path)
-    mode = black_config or Mode()
+    encoding: Encoding = "utf-8",
+) -> FileContent:
+    """
+    Format arbitrary bytes for the given path.
+
+    Requires passing valid config objects for both black and µsort. If the given path
+    represents a type stub (has a ``.pyi`` suffix), the black config object will be
+    updated to set ``is_pyi = True``.
+
+    Note: content will be decoded before passing to black, and re-encoded to bytes
+    again. Be sure to pass a known-valid unicode encoding for the content.
+    :func:`ufmt.util.read_file` can be used to both read bytes from disk, and make a
+    best guess at file encodings.
+    """
+    result = usort(content, usort_config, path)
 
     if path.suffix == ".pyi":
-        mode.is_pyi = True
+        black_config.is_pyi = True
```

Comment:
I don't think you meant to alter the user-passed object like this?

Reply:
It actually maintains the same-ish behavior as before, though I guess the better option would be to just duplicate/evolve the config object with `is_pyi=True`.